packages/flutter_tools/lib/src/build_runner/web_compilation_delegate.dart

```diff
@@ -197,3 +204,81 @@ class BuildRunnerWebCompilationProxy extends WebCompilationProxy {
     }
   }
 }
+
+/// Handles mapping a single root file scheme to a multiroot scheme.
+///
+/// This allows one build_runner build to read the output from a previous
+/// isolated build.
+class MultirootFileBasedAssetReader extends core.FileBasedAssetReader {
+  MultirootFileBasedAssetReader(
+    core.PackageGraph packageGraph,
+    this.generatedDirectory,
+  ) : super(packageGraph);
+
+  final Directory generatedDirectory;
+
+  @override
+  Future<bool> canRead(AssetId id) {
+    if (packageGraph[id.package] == packageGraph.root &&
+      !fs.file(path.join(packageGraph.root.path, id.path)).existsSync()) {
+        return fs.file(
+          path.join(generatedDirectory.path, packageGraph.root.name, id.path)
+        ).exists();
+    }
+    return super.canRead(id);
+  }
+
+  @override
+  Future<List<int>> readAsBytes(AssetId id) {
+    if (packageGraph[id.package] == packageGraph.root &&
+      !fs.file(path.join(packageGraph.root.path, id.path)).existsSync()) {
+        return fs.file(
+          path.join(generatedDirectory.path, packageGraph.root.name, id.path)
+        ).readAsBytes();
+    }
+    return super.readAsBytes(id);
+  }
+
+  @override
+  Future<String> readAsString(AssetId id, {Encoding encoding}) {
+    if (packageGraph[id.package] == packageGraph.root &&
+      !fs.file(path.join(packageGraph.root.path, id.path)).existsSync()) {
+        return fs.file(
+          path.join(generatedDirectory.path, packageGraph.root.name, id.path)
+        ).readAsString();
+    }
+    return super.readAsString(id, encoding: encoding);
+  }
+
+  @override
+  Stream<AssetId> findAssets(Glob glob, {String package}) async* {
+    if (package == null || packageGraph.root.name == package) {
+      await for (io.FileSystemEntity  entity in glob.list(followLinks: true, root: packageGraph.root.path)) {
+        if (entity is io.File && !path.basename(entity.path).startsWith('._')) {
+          yield _fileToAssetId(entity, packageGraph.root);
+        }
+      }
+      final String generatedRoot = fs.path.join(
+        generatedDirectory.path, packageGraph.root.name
+      );
+      if (!fs.isDirectorySync(generatedRoot)) {
+        return;
+      }
+      await for (io.FileSystemEntity entity in glob.list(followLinks: true, root: generatedRoot)) {
+        if (entity is io.File && !path.basename(entity.path).startsWith('._')) {
```

Comment:
Could you extract this expression into a named function so it's clear what it does? E.g. `_isNonHiddenFile(entity)` or sth (I'm not actually sure what those files are).